src/js/components/FrameworkConfigurationForm.js

```diff
@@ -0,0 +1,308 @@
+    this.props.onFormDataChange(formData);
+    this.setState({ errorSchema });
+  }
+
+  validate(formData, errors) {
+    const { packageDetails } = this.props;
+
+    this.writeErrors(formData, packageDetails.getConfig(), false, errors);
+
+    return errors;
+  }
+
+  formatErrorForJSON(errorSchema, path) {
+    if (!errorSchema.__errors) {
+      return [];
+    }
+    const currentLevelErrors = errorSchema.__errors.map(message => {
+      return { message, path };
+    });
+
+    return Object.keys(errorSchema)
+      .map(key => this.formatErrorForJSON(errorSchema[key], path.concat(key)))
+      .reduce((a, b) => a.concat(b), currentLevelErrors);
+  }
+
+  getErrorsForJSONEditor() {
+    const { errorSchema } = this.state;
+
+    if (!errorSchema) {
+      return [];
+    }
+
+    return this.formatErrorForJSON(errorSchema, []);
+  }
+
+  render() {
+    const {
+      packageDetails,
+      jsonEditorActive,
+      formData,
+      activeTab,
+      deployErrors
+    } = this.props;
+    const schema = packageDetails.getConfig();
+    const jsonEditorErrors = this.getErrorsForJSONEditor();
+
+    const TitleField = props => {
+      return (
+        <DefaultTitleField
+          {...props}
+          title={this.getFormattedSectionLabel(props.title)}
+          required={false}
+        />
+      );
+    };
+
+    const jsonEditorPlaceholderClasses = classNames(
+      "modal-full-screen-side-panel-placeholder",
+      { "is-visible": jsonEditorActive }
+    );
+    const jsonEditorClasses = classNames("modal-full-screen-side-panel", {
+      "is-visible": jsonEditorActive
+    });
+
+    let errorsAlert = null;
+    if (deployErrors) {
+      errorsAlert = <CosmosErrorMessage error={deployErrors} />;
+    }
+
+    return (
+      <div className="flex flex-item-grow-1">
+        <div className="create-service-modal-form__scrollbar-container modal-body-offset gm-scrollbar-container-flex">
+          <PageHeaderNavigationDropdown
+            handleNavigationItemSelection={this.handleDropdownNavigationSelection.bind(
+              this
+            )}
+            items={this.getDropdownNavigationList()}
+          />
+          <div className="modal-body-padding-surrogate create-service-modal-form-container">
+            <div className="create-service-modal-form container container-wide">
+              <Tabs
+                activeTab={activeTab}
+                handleTabChange={this.handleTabChange.bind(this)}
+                vertical={true}
+                className={"menu-tabbed-container-fixed"}
+              >
+                <TabButtonList>
+                  {this.getFormTabList()}
+                </TabButtonList>
+                <div className="menu-tabbed-view-container">
+                  {errorsAlert}
+                  <SchemaForm
+                    schema={schema}
+                    formData={formData}
+                    onChange={this.handleFormChange.bind(this)}
```

Comment:
This will always create a new function which might mess up the rerendering as the props are changing. To omit this we could bind this to the function in the constructor

```JS
this.handleFormChange = this.handleFormChange.bind(this);
```

This also applies for the `validate` prop.

Reply:
Yeah, I was going to change all these to bind in the constructor! Thanks for reminder.